doc/man/man4/npi_conv.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/npi_conv.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH NPI_CONV 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Modules"
.\"
.\"
.SH NAME
.B npi_conv
\- \fISTREAMS\fR NPI endian conversion module
.\"
.\"
.SH SYNOPSIS
.PP
.HP 8
.BI "int " fd ;
.PD 0
.HP
.BI "int open(" \*(lq\fR/dev/nuls\fP\*(rq ", " \fR0\fP );
.HP
.BI "int ioctl(" fd ", " I_PUSH ", " \*(lqnpi_conv\*(rq );
.PD
.\"
.\"
.SH DESCRIPTION
.PP
.B npi_conv
is a
.I STREAMS
module that is pushed over any
.IR Stream ,
.IR STREAMS -based
.BR pipe (4)
or
.IR STREAMS -based
.BR fifo (4s).
.PP
Aside from being a
.I STREAMS
module,
.B npi_conv
performs no actions other than to convert
.IR "Network Provider Interface" ,
.[
npi revision 2.0.0 draft 2
.]
.BR npi (7),
messages received from above into network-byte-ordered
messages passed downward (resp.); and visa versa in the upstream direction.
That is, each
.BR npi (7)
message received from above is changed to a
.BR npi (7)
message with network-byte-ordering and passed downstream without queuing nor consideration for flow
control.  Also, any
.BR npi (7)
message received from below is converted to a
.BR npi (7)
message with host-byte-ordering and passed upstream.
.\"
.\"
.SH USAGE
.PP
The primary purpose of the
.B npi_conv
module is to translate between
.IR "Network Provider Interface" ,
.[
npi revision 2.0.0 draft 2
.]
messages in host and network byte ordering.  The module is simply pushed over
any driver that implements the
.BR npi (7)
protocol in network byte order, or that otherwise requires conversion to and
from host and network byte ordering for
.BR npi (7)
messages.
.\"
.\"
.SH NOTICES
.PP
.BR npi_conv
is implementation-specific to the
.B @PACKAGE_TITLE@
package.
.\"
.\"
.\".SH EXAMPLES
.\".PP
.\"
.\"
.SH MODULES
.PP
.B npi_conv
is a
.I STREAMS
module named
.IR \*(lqnpi_conv\*(rq .
.\"
.\"
.SH "SEE ALSO"
.PP
.BR npi (7),
.BR open (2s),
.BR ioctl (2s),
.BR streamio (7),
.BR putnext (9),
.BR M_PROTO (9),
.BR M_PCPROTO (9),
.BR M_CTL (9),
.BR M_PCCTL (9).
.\"
.\"
.SH BUGS
.PP
.B npi_conv
has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
.B npi_conv
is compatible with the modules implementing the
.IR "Network Provider Interface" ,
.[
npi revision 2.0.0 draft 2
.]
.BR npi (7),
or in any case where it is necessary to convert
.BR npi (7)
messages to and from host and network byte ordering across the module.
.IP \(em \w'\(em\(em'u
.B npi_conv
is specific to
.BR @PACKAGE_TITLE@ ,
and will not be used by portable
.I STREAMS
modules, drivers and applications programs.
.IP \(em
The name
.B npi_conv
is not standard and was chosen for
.BR @PACKAGE_TITLE@ .
.IP \(em
.B npi_conv
is used primarily for passing
.BR npi (7)
primitive directly across a network connection.  It will not be used by
portable
.I STREAMS
modules, drivers and applications.
.PP
See
.BR STREAMS (9)
for additional compatibility information.
.\"
.\"
.SH CONFORMANCE
.PP
.B npi_conv
is implementation-specific and is not subject to standardization.
.\"
.\"
.SH HISTORY
.PP
.B npi_conv
first appeared in
.B @PACKAGE_TITLE@
in release
.BR openss7-1.1.1-20101117 .
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/modules/npi_conv.c
/*****************************************************************************

 @(#) File: src/modules/npi_conv.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/modules/npi_conv.c (" PACKAGE_ENVR ") " PACKAGE_DATE;


/*
 *  NPI-CONV is a simple endian conversion module for use with the RMUX driver.  It converts service
 *  primitives fron host to network byte order and visa versa.  It is pushed over a NPI Stream to
 *  perform the conversion.  Because the module neither allocates nor frees messages, no service
 *  procedure is required and the conversion is performed completely in the put procedure.
 */

#define _SUN_SOURCE		1

#include <sys/os7/compat.h>
#include <sys/npi.h>

#define NPI_CONV_DESCRIP	"NPI Endian Conversion (NPI-CONV) for Linux Fast-STREAMS"
#define NPI_CONV_EXTRA		"Part of Linux Fast-STREAMS"
#define NPI_CONV_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define NPI_CONV_REVISION	"OpenSS7 src/modules/npi_conv.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define NPI_CONV_DEVICE		"SVR 4.2 NPI Endian Conversion (NPI-CONV) for STREAMS"
#define NPI_CONV_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define NPI_CONV_LICENSE	"GPL"
#define NPI_CONV_BANNER		NPI_CONV_DESCRIP	"\n" \
				NPI_CONV_EXTRA		"\n" \
				NPI_CONV_COPYRIGHT	"\n" \
				NPI_CONV_REVISION	"\n" \
				NPI_CONV_DEVICE		"\n" \
				NPI_CONV_CONTACT	"\n"
#define NPI_CONV_SPLASH		NPI_CONV_DEVICE		" - " \
				NPI_CONV_REVISION	"\n"

#ifdef MODULE
MODULE_AUTHOR(NPI_CONV_CONTACT);
MODULE_DESCRIPTION(NPI_CONV_DESCRIP);
MODULE_SUPPORTED_DEVICE(NPI_CONV_DEVICE);
MODULE_LICENSE(NPI_CONV_LICENSE);
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-npi-conv");
#endif				/* MODULE_ALIAS */
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif
#endif				/* MODULE */

#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-modid-" __stringify(CONFIG_STREAMS_NPI_CONV_MODID));
MODULE_ALIAS("streams-module-npi-conv");
#endif				/* MODULE_ALIAS */

/*
 * The point of this module is only to convert M_PROTO messages to and from network byte order.  All
 * other messages are passed along.  It is the responsibility of the upper layer module to convert
 * other message types.
 */

static int
npi_m_proto(queue_t *q, mblk_t *mp, uint32_t prim)
{
	union NPI_primitives *p = (typeof(p)) mp->b_rptr;

	/* Note that addresses must be specified in network byte order for this to work. */
	switch (prim) {
	case N_CONN_REQ:
	case N_CONN_RES:
	case N_DISCON_REQ:
	case N_DATA_REQ:
	case N_EXDATA_REQ:
	case N_INFO_REQ:
	case N_BIND_REQ:
	case N_UNBIND_REQ:
	case N_UNITDATA_REQ:
	case N_OPTMGMT_REQ:
	case N_CONN_IND:
	case N_CONN_CON:
	case N_DISCON_IND:
	case N_DATA_IND:
	case N_EXDATA_IND:
	case N_INFO_ACK:
	case N_BIND_ACK:
	case N_ERROR_ACK:
	case N_OK_ACK:
	case N_UNITDATA_IND:
	case N_UDERROR_IND:
	case N_DATACK_REQ:
	case N_DATACK_IND:
	case N_RESET_REQ:
	case N_RESET_IND:
	case N_RESET_RES:
	case N_RESET_CON:
	default:
		break;
	}
	putnext(q, mp);
	return (0);
}

static streamscall int
npi_rput(queue_t *q, mblk_t *mp)
{
	if (datamsg(DB_TYPE(mp))) {
		switch (DB_TYPE(mp)) {
		case M_DATA:
			break;
		case M_PROTO:
		case M_PCPROTO:
			if (mp->b_wptr >= mp->b_rptr + sizeof(uint32_t)) {
				uint32_t prim;

				prim = *(uint32_t *) mp->b_rptr;
				*(uint32_t *) mp->b_rptr = ntohl(prim);
				return npi_m_proto(q, mp, prim);
			}
			break;
		default:
		case M_DELAY:
			break;
		}
	}
	putnext(q, mp);
	return (0);
}

static streamscall int
npi_wput(queue_t *q, mblk_t *mp)
{
	if (datamsg(DB_TYPE(mp))) {
		switch (DB_TYPE(mp)) {
		case M_DATA:
			break;
		case M_PROTO:
		case M_PCPROTO:
			if (mp->b_wptr >= mp->b_rptr + sizeof(uint32_t)) {
				uint32_t prim;

				prim = htonl(*(uint32_t *) mp->b_rptr);
				*(uint32_t *) mp->b_rptr = prim;
				return npi_m_proto(q, mp, prim);
			}
			break;
		default:
		case M_DELAY:
			break;
		}
	}
	putnext(q, mp);
	return (0);
}

static streamscall int
npi_qopen(queue_t *q, dev_t *devp, int oflags, int sflag, cred_t *crp)
{
	if (q->q_ptr != NULL)
		return (0);
	if (sflag != MODOPEN || !q->q_next)
		return (EIO);
	q->q_ptr = WR(q)->q_ptr = (queue_t *) 1;	/* just mark it open */
	qprocson(q);
	return (0);
}

modID_t modid = CONFIG_STREAMS_NPI_CONV_MODID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else				/* module_param */
module_param(modid, ushort, 0444);
#endif				/* module_param */
MODULE_PARM_DESC(modid, "Module ID for NPI-CONV module.");

static struct module_info npi_conv_minfo = {
	.mi_idnum = CONFIG_STREAMS_NPI_CONV_MODID,
	.mi_idname = CONFIG_STREAMS_NPI_CONV_NAME,
	.mi_minpsz = STRMINPSZ,
	.mi_maxpsz = STRMAXPSZ,
	.mi_hiwat = STRHIGH,
	.mi_lowat = STRLOW,
};

static struct module_stat npi_conv_rstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));
static struct module_stat npi_conv_wstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));

static streamscall int
npi_qclose(queue_t *q, int oflag, cred_t *credp)
{
	qprocsoff(q);
	q->q_ptr = WR(q)->q_ptr = (queue_t *) 0;	/* just mark it closed */
	return (0);
}

static struct qinit npi_rinit = {
	.qi_putp = npi_rput,
	.qi_srvp = NULL,
	.qi_qopen = npi_qopen,
	.qi_qclose = npi_qclose,
	.qi_minfo = &npi_conv_minfo,
	.qi_mstat = &npi_conv_rstat,
};

static struct qinit npi_winit = {
	.qi_putp = npi_wput,
	.qi_srvp = NULL,
	.qi_minfo = &npi_conv_minfo,
	.qi_mstat = &npi_conv_wstat,
};

struct streamtab npi_convinfo = {
	.st_rdinit = &npi_rinit,
	.st_wrinit = &npi_winit,
};

static struct fmodsw npi_fmod = {
	.f_name = CONFIG_STREAMS_NPI_CONV_NAME,
	.f_str = &npi_convinfo,
	.f_flag = D_MP,
	.f_kmod = THIS_MODULE,
};

int __init
npi_convinit(void)
{
	int err;

#ifdef CONFIG_STREAMS_NPI_CONV_MODULE
	cmn_err(CE_NOTE, NPI_CONV_BANNER);
#else				/* CONFIG_STREAMS_NPI_CONV_MODULE */
	cmn_err(CE_NOTE, NPI_CONV_SPLASH);
#endif				/* CONFIG_STREAMS_NPI_CONV_MODULE */
	npi_conv_minfo.mi_idnum = modid;
	if ((err = register_strmod(&npi_fmod)) < 0)
		return (err);
	if (modid == 0 && err > 0)
		modid = err;
	return (0);
}

void __exit
npi_convexit(void)
{
	unregister_strmod(&npi_fmod);
}

#ifdef CONFIG_STREAMS_NPI_CONV_MODULE
module_init(npi_convinit);
module_exit(npi_convexit);
#endif				/* CONFIG_STREAMS_NPI_CONV_MODULE */
